### DISCUSSION TITLE
Avoid PacerLoginException when calling recap-fetch?

### DISCUSSION THREAD

**Discussion by stephen-jones-praedicat** [freelawproject/courtlistener#3994] [Q&A]
About 40% of our recap-fetch requests are resulting in PacerLoginException errors. We make the calls again and eventually they all work. Is there something we could be doing differently to avoid running into this problem?

I found this but I'm not sure whether it's related: https://github.com/freelawproject/courtlistener/issues/3423#issuecomment-1982208356

Any help would be appreciated.

Thanks,

Stephen

**Comment by mlissner** [comment]
Hm, are you sending a lot of requests in around the same time? I think that can be part of the issue. I think we need more sophistication around what we do when credentials time out, and I think what can happen is that when they time out, if there are a lot of pending items, many of those will fail before we can log back in with a future request. 

Maybe try spacing requests out by 5s or so, and see if that could help?

**Reply by stephen-jones-praedicat** [reply]
When we have a queue of items to send to recap-fetch we aren't pausing at all. I'll make that change right away 

Thank you, that's exactly the kind of tip I was hoping for. I'll let you know if it makes a difference.

**Reply by stephen-jones-praedicat** [reply]
It improves it a bit but we're still getting a lot of PacerLoginExceptions. Is there any point in me spacing them a bit more?

**Reply by mlissner** [reply]
It's worth a try, yes. This is definitely an area where we could do a better job for you though. 

It's difficult to handle this properly because one of our goals is to throw away your credentials after logging you in. That means that when the cookies time out, we can't fix the problem until we get your credentials again, and I don't think we connect the dots between:

 - The credentials just failed

And:

 - Therefore, we should log in the next opportunity we get. 

Instead, we cache your cookie for a while and don't clear it out until it times out, even if it starts failing. 

This is all from memory, so I might be remembering incorrectly, but I am aware we're not nailing this feature so that you can avoid retries.
